pkg/reconciler/knativeserving/knativeserving_controller.go

```diff
@@ -295,5 +300,26 @@ func (r *Reconciler) deleteObsoleteResources(manifest *mf.Manifest, instance *se
 	if err := manifest.Delete(resource, &metav1.DeleteOptions{}); err != nil {
 		return err
 	}
+
+	// If there used to be a CR at an older version, we need to remove the old resources with
+	// the label serving.knative.dev/release="<oldVersion>"
+	if r.oldVersion != "" && r.oldVersion != version.Version {
+		selector := fmt.Sprintf("%s=%s", selectorKey, r.oldVersion)
+		listOptions := metav1.ListOptions{ LabelSelector: selector }
+
+		// Remove all the deployments at the old version
+		deploymentList, err := r.KubeClientSet.AppsV1().Deployments(instance.GetNamespace()).List(listOptions)
+		if err == nil {
+			for _, dep := range deploymentList.Items {
+				if e := r.KubeClientSet.AppsV1().Deployments(instance.GetNamespace()).Delete(dep.Name,
```

Comment:
This makes me nervous. Deleting a deployment instead of updating it introduces the potential for losing requests, I think. Besides, I prefer seeing in this function exactly which resources are obsolete and need to be deleted.

Reply:
@jcrossley3 To find out all the resources to be removed, I can think of using the manifestival package to read the manifest of the older version, and compare to the manifest of the new version, to see if they still exist in the new version. Do you have any other suggestions?